### DISCUSSION TITLE
Cross platform builds?

### DISCUSSION THREAD

**Discussion by arnaudvergnet** [organicmaps/organicmaps#1865] [Q&A]
Do not take this thread as a request, I am just curious.

Qt can build for desktop, Android and iOS, so I thought using one codebase for 3 platforms would be easier to maintain. I worked with other frameworks like React Native, it helped me speed up development and reduce maintenance by quite a lot. But I never used Qt for mobile apps so I do not really know its limitations.

Do you know if it would be technically possible? I tried the desktop version and I can feel it has been left out and is nowhere near the Android and iOS versions. So even if it was technically possible, I know it would mean a lot of work.

What made me consider this idea is the emergence of new mainline Linux mobile OSes like [KDE Plasma Mobile](https://plasma-mobile.org). Plasma uses the [Maui Kit](https://mauikit.org/) which is a cross platform Qt framework for building convergent applications. Such platforms are in need of a map application like Organic Maps. If Om's Qt version was compatible with mobile, targeting those new Linux platforms would be easy and bring new developers onboard. 

What are your thoughts on this?

**Comment by biodranik** [comment]
QT is very bad for iOS and Android. But it may work for Linux, especially if it's already QT-based. So it looks like:
1. Somehow split desktop version from mobile Linux version (we need to keep a desktop version at least for development and for debugging). The best way could be to start a new Qt project and reuse the core app's libraries.
2. Design a new mobile Linux UI and implement it.
3. Add support of some system-dependent APIs if necessary (location, touch screen, etc.)

**Reply by arnaudvergnet** [reply]
And what about other cross platform frameworks like React Native and Flutter, did you ever consider those? Compared to when Maps.me started, they are now a real alternative to native apps.

**Reply by biodranik** [reply]
Looks like there are some C++ bindings for Flutter and React Native. So in theory it would be possible. Do you know any good cross-platform projects using them?

I'm personally very skeptical about the battery life for both mentioned frameworks. My bet is that they always use more (or way more) battery than a properly written native platform code. Saving a battery is a crucial part of offline maps, especially for longer journeys.

**Reply by arnaudvergnet** [reply]
I don't know any app making use of cpp bindings directly, but I know some libraries. For React Native, you have [mapbox-gl](https://github.com/react-native-mapbox-gl/maps) making use of the [MapLibre](https://github.com/maplibre/maplibre-gl-native) native library which uses cpp bindings. The mapbox-gl component simply provides bindings between javascript and native.

So my guess is that you will have to make basic android and iOS components interacting with the cpp and acting as backend, and use Flutter or React Native only as a common frontend for Android and iOS.

But OM seems to have a lot of cpp bindings, so I don't know if it would speed up the development. And this approach would add some overhead I did not measure, but it would surely impact power consumption compared to native.

Thanks for the information, I was curious but now I understand cross-platform builds would not be optimal for your goal.

**Reply by biodranik** [reply]
The cross-platform approach always failed in my experience. Building a clean native Linux UI would be the best, IMHO.

**Comment by arnaudvergnet** [comment]
> From what I can see, that's not quite true, at least in theory - Qt Quick Controls support [styles](https://doc.qt.io/qt-6/qtquickcontrols-styles.html), and [Material Style](https://doc.qt.io/qt-6/qtquickcontrols-material.html) based on the Google Material Design Guidelines is one of those.

@bam80 Let's continue our discussion here.

UI is one thing, there's also UX. For example Android, iOS and Kirigami all have their own headers with different behaviors. Some UX patterns might be normal on one platform but not on another.

Once the OM Qt app is compatible with mobile, it would be great to do some testing! Having to deal with only one framework for all platforms would speed up development quite a lot.

**Comment by trymeouteh** [comment]
Would like to see a flutter version. Flutter allows you to compile apps for desktop and mobile. I think long term it will make development easier when making releases for each OS.
